
# Stochastic Demography
## Excercise3

- <a href='#1'>1</a>
- <a href='#2'>2</a>
- <a href='#3'>3</a>
- <a href='#4'>4</a>
- <a href='#5'>5</a>
- <a href='#6'>6</a>
- <a href='#7'>7</a>
- <a href='#8'>8</a>
- <a href='#9'>9</a>


In [2]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 

In [3]:
df = pd.read_excel('/home/oscar/Dropbox/StochasticDemography/data/life_table_fin2015_male_female.xlsx')
df = df.iloc[:, :10]

df_male = df.iloc[250:362,:]
df_female = df.iloc[136:247, :]
df_male.columns = df.iloc[249,:]
df_female.columns = df.iloc[135, :]
df_male.columns.set_names('male',inplace=True)
df_female.columns.set_names('female',inplace=True)
df_female.index = range(len(df_female))
df_male.index = range(len(df_male))

In [53]:
df_male.shape
df_female.shape

df_female
df_male


(111, 10)

(111, 10)

female  Year age x     m(x)     q(x)  a(x)    l(x)  d(x)   L(x)     T(x)  \
0       2015     0  0.00178  0.00178  0.15  100000   178  99848  8419059   
1       2015     1    4e-05    4e-05   0.5   99822     4  99821  8319211   
2       2015     2    3e-05    3e-05   0.5   99819     3  99817  8219390   
3       2015     3   0.0001   0.0001   0.5   99815    10  99810  8119573   
4       2015     4    7e-05    7e-05   0.5   99805     7  99802  8019763   
5       2015     5    7e-05    7e-05   0.5   99799     7  99795  7919961   
6       2015     6        0        0   0.5   99792     0  99792  7820165   
7       2015     7    7e-05    7e-05   0.5   99792     7  99789  7720373   
8       2015     8    3e-05    3e-05   0.5   99786     3  99784  7620584   
9       2015     9   0.0001   0.0001   0.5   99782    10  99777  7520800   
10      2015    10    7e-05    7e-05   0.5   99772     7  99769  7421023   
11      2015    11  0.00014  0.00014   0.5   99765    14  99758  7321255   
12      2015    12  0.00014  0.00014   0.5   99751    14  99744  7221496   
13      2015    13    7e-05    7e-05   0.5   99737     7  99734  7121752   
14      2015    14    7e-05    7e-05   0.5   99730     7  99727  7022018   
15      2015    15  0.00014  0.00014   0.5   99723    14  99716  6922291   
16      2015    16  0.00014  0.00014   0.5   99709    14  99703  6822575   
17      2015    17  0.00021  0.00021   0.5   99696    21  99685  6722872   
18      2015    18  0.00023  0.00023   0.5   99675    23  99664  6623187   
19      2015    19  0.00032  0.00032   0.5   99652    32  99636  6523523   
20      2015    20  0.00012  0.00012   0.5   99620    12  99614  6423887   
21      2015    21  0.00018  0.00018   0.5   99608    18  99599  6324273   
22      2015    22   0.0003   0.0003   0.5   99590    30  99575  6224674   
23      2015    23  0.00021  0.00021   0.5   99560    21  99550  6125099   
24      2015    24  0.00012  0.00012   0.5   99539    12  99534  6025550   
25      2015    25  0.00021  0.00021   0.5   99528    20  99518  5926016   
26      2015    26  0.00018  0.00018   0.5   99507    18  99498  5826498   
27      2015    27  0.00021  0.00021   0.5   99490    21  99479  5727000   
28      2015    28  0.00031  0.00031   0.5   99468    31  99453  5627521   
29      2015    29  0.00033  0.00033   0.5   99438    33  99421  5528068   
..       ...   ...      ...      ...   ...     ...   ...    ...      ...   
81      2015    81  0.03874    0.038   0.5   71622  2722  70261   651956   
82      2015    82  0.04769  0.04658   0.5   68900  3209  67295   581695   
83      2015    83  0.05509  0.05361   0.5   65690  3522  63930   514400   
84      2015    84  0.06365  0.06168   0.5   62169  3835  60251   450470   
85      2015    85  0.07532  0.07259   0.5   58334  4234  56217   390219   
86      2015    86  0.09169  0.08767   0.5   54100  4743  51728   334002   
87      2015    87   0.0956  0.09124   0.5   49357  4503  47105   282274   
88      2015    88  0.11129  0.10542   0.5   44853  4729  42489   235169   
89      2015    89  0.13191  0.12375   0.5   40125  4965  37642   192680   
90      2015    90  0.15421  0.14317   0.5   35159  5034  32642   155038   
91      2015    91  0.17391  0.15999   0.5   30126  4820  27716   122396   
92      2015    92  0.18814  0.17196   0.5   25306  4352  23130    94680   
93      2015    93  0.22491  0.20217   0.5   20954  4236  18836    71550   
94      2015    94  0.23981  0.21413   0.5   16718  3580  14928    52714   
95      2015    95  0.27978  0.24544   0.5   13138  3225  11526    37787   
96      2015    96  0.31264  0.27038   0.5    9913  2680   8573    26261   
97      2015    97   0.3475  0.29606   0.5    7233  2141   6162    17688   
98      2015    98  0.38408   0.3222   0.5    5092  1641   4271    11525   
99      2015    99  0.42202  0.34848   0.5    3451  1203   2850     7254   
100     2015   100   0.4609  0.37458   0.5    2248   842   1827     4404   
101     2015   101  0.50026  0.40017   0.5    14

male  Year age x     m(x)     q(x)  a(x)    l(x)  d(x)   L(x)     T(x)   e(x)
0     2015     0  0.00159  0.00159  0.15  100000   159  99864  7862257  78.62
1     2015     1  0.00013  0.00013   0.5   99841    13  99835  7762393  77.75
2     2015     2  0.00023  0.00023   0.5   99828    23  99816  7662558  76.76
3     2015     3    6e-05    6e-05   0.5   99805     6  99802  7562742  75.78
4     2015     4    6e-05    6e-05   0.5   99799     6  99795  7462940  74.78
5     2015     5    9e-05    9e-05   0.5   99792     9  99788  7363145  73.78
6     2015     6    3e-05    3e-05   0.5   99783     3  99781  7263357  72.79
7     2015     7    3e-05    3e-05   0.5   99780     3  99778  7163576  71.79
8     2015     8    3e-05    3e-05   0.5   99776     3  99775  7063798   70.8
9     2015     9    6e-05    6e-05   0.5   99773     6  99770  6964023   69.8
10    2015    10    3e-05    3e-05   0.5   99767     3  99765  6864253   68.8
11    2015    11   0.0002   0.0002   0.5   99764    20  99754  6764488  67.81
12    2015    12    3e-05    3e-05   0.5   99744     3  99742  6664734  66.82
13    2015    13   0.0002   0.0002   0.5   99740    20  99730  6564992  65.82
14    2015    14    3e-05    3e-05   0.5   99720     3  99718  6465262  64.83
15    2015    15   0.0002   0.0002   0.5   99717    20  99707  6365544  63.84
16    2015    16  0.00023  0.00023   0.5   99697    23  99685  6265837  62.85
17    2015    17   0.0002   0.0002   0.5   99674    20  99664  6166152  61.86
18    2015    18  0.00035  0.00035   0.5   99654    35  99637  6066488  60.88
19    2015    19  0.00064  0.00064   0.5   99619    64  99587  5966851   59.9
20    2015    20  0.00068  0.00068   0.5   99555    67  99522  5867264  58.93
21    2015    21  0.00079  0.00079   0.5   99488    79  99448  5767742  57.97
22    2015    22   0.0008  0.00079   0.5   99409    79  99369  5668294  57.02
23    2015    23  0.00056  0.00056   0.5   99330    56  99302  5568924  56.06
24    2015    24  0.00096  0.00095   0.5   99274    95  99227  5469622   55.1
25    2015    25  0.00087  0.00087   0.5   99179    86  99136  5370396  54.15
26    2015    26  0.00088  0.00088   0.5   99093    88  99049  5271260   53.2
27    2015    27   0.0006   0.0006   0.5   99005    60  98976  5172210  52.24
28    2015    28  0.00067  0.00067   0.5   98946    67  98912  5073235  51.27
29    2015    29  0.00094  0.00094   0.5   98879    93  98832  4974323  50.31
..     ...   ...      ...      ...   ...     ...   ...    ...      ...    ...
81    2015    81  0.06254  0.06064   0.5   52731  3198  51132   399912   7.58
82    2015    82  0.07083   0.0684   0.5   49533  3388  47839   348780   7.04
83    2015    83  0.08445  0.08103   0.5   46145  3739  44275   300941   6.52
84    2015    84  0.09187  0.08784   0.5   42406  3725  40543   256666   6.05
85    2015    85  0.11038  0.10461   0.5   38681  4046  36658   216122   5.59
86    2015    86  0.12955  0.12167   0.5   34635  4214  32528   179465   5.18
87    2015    87  0.14081  0.13155   0.5   30421  4002  28420   146937   4.83
88    2015    88  0.15258  0.14177   0.5   26419  3745  24546   118517   4.49
89    2015    89  0.17573  0.16154   0.5   22673  3663  20842    93971   4.14
90    2015    90  0.19554  0.17812   0.5   19011  3386  17318    73129   3.85
91    2015    91  0.20226  0.18369   0.5   15625  2870  14190    55811   3.57
92    2015    92  0.23207  0.20794   0.5   12755  2652  11429    41622   3.26
93    2015    93  0.28603  0.25024   0.5   10102  2528   8838    30193   2.99
94    2015    94  0.28684  0.25086   0.5    7574  1900   6624    21355   2.82
95    2015    95  0.32743  0.28136   0.5    5674  1597   4876    14731    2.6
96    2015    96  0.35892  0.30431   0.5    4078  1241   3457     9855   2.42
97    2015    97  0.39167  0.32753   0.5    2837   929   2372     6397   2.26
98    2015    98  0.42543  0.35081   0.5    1908   669   1573     4025   2.11
99    2015    99   0.4599  0.37392   0.5    1238   463   1007     2452   1.98
100   2015   100  0.49

## 1
<a id='1'></a>
Consider again the grandmother aged 55, and the grandson, aged 5, from exercise 4 last
week. What is the probability that the grandmother survives her grandson?

$A_f$ death time of grandmother

$A_m$ death time of grandson

$$\sum_{k=0}^{54} P(A_{f} > k + 55 \mid A_{f}>54+k)P(A_{m}=5+k \mid A_{m}>4+k) \\=  
\sum_{k=0}^{54} \frac{l_{56 + k}^{(f)}}{l_{55 + k}^{(f)}} (1 - \frac{l_{6 + k}^{(m)}}{l_{5+k}^{(m)}})$$


In [68]:
lf = df_female['l(x)'].values
lm = df_male['l(x)'].values

p = 0
af = 55
am = 5

for k in range(110- ag):
    p += lf[af + 1 + k]/lf[af + k] * (1 - lm[am + 1 + k] / lm[am + k])

p



0.06419380751862737

## 2 
<a id='2'></a>
Discuss the following statement: For the minorities in the United States, those that survive middle age, live longer than their white counterparts.

ANSWER:

- Hispanic: 

  - all: $e_{40} = 42.7$ years
  
  - female: $e_{40} = 44.5$ years
  
  - male: $e_{40} = 40.5$ years
  

- White: 

    - all: $e_{40} = 40.3$ years

    - female: $e_{40} = 42.2$ years
    
    - male: $e_{40} = 38.3$ years


- Black:

    - all: $e_{40} = 37.0$ years
    
    - female: $e_{40} = 39.4$ years
    
    - male: $e_{40} = 34.2$ years

Hispanics have longest expected remaining life time at year 40.

## 3
<a id='3'></a>
Discuss the following statement: Those born in 1918 and survived childhood, lived longer than their counterparts born in 1916 or 1920.

ANSWER:

- Born in 1918 
   - $e_{1} = 60,34$
- Born in 1916
   - $e_{1} = 58,56$
- Born in 1920
   - $e_{1} = 62,36$


In [62]:
df_3 = pd.read_excel('/home/oscar/Dropbox/StochasticDemography/data/FinlandMortCohorts 1912-22.xlsx', skiprows=2)

In [63]:
df_16 = df_3[df_3['Year'] == 1916]

df_18 = df_3[df_3['Year'] == 1918]

df_20 = df_3[df_3['Year'] == 1920]

df = pd.DataFrame({1916: df_16['ex'].values, 
                   1918: df_18['ex'].values,
                   1920: df_20['ex'].values,
                  })
df.head(50)


1916   1918   1920
0   52.60  53.38  57.16
1   58.56  60.34  62.36
2   59.87  62.39  62.98
3   61.05  62.43  62.84
4   61.28  61.95  62.44
5   60.73  61.28  61.76
6   60.07  60.60  61.00
7   59.29  59.83  60.21
8   58.49  59.02  59.39
9   57.70  58.20  58.57
10  56.83  57.36  57.72
11  55.99  56.51  56.88
12  55.12  55.67  56.03
13  54.28  54.83  55.18
14  53.46  53.98  54.33
15  52.65  53.13  53.48
16  51.83  52.30  52.64
17  51.03  51.48  51.83
18  50.21  50.70  51.04
19  49.45  49.91  50.25
20  48.67  49.13  49.63
21  47.92  48.36  49.48
22  47.17  48.45  50.23
23  46.43  48.35  49.85
24  46.99  48.72  49.69
25  46.73  48.28  49.48
26  46.90  47.94  48.75
27  46.35  47.51  48.00
28  45.91  46.75  47.22
29  45.43  45.97  46.42
30  44.67  45.17  45.61
31  43.87  44.36  44.74
32  43.06  43.52  43.85
33  42.25  42.66  42.96
34  41.38  41.78  42.06
35  40.52  40.88  41.15
36  39.63  39.98  40.25
37  38.72  39.08  39.36
38  37.84  38.21  38.48
39  36.95  37.32  37.58
40  36.06  36.45  36.70
41  35.18  35.55  35.81
42  34.29  34.66  34.94
43  33.41  33.78  34.07
44  32.53  32.91  33.22
45  31.67  32.07  32.38
46  30.83  31.22  31.55
47  30.00  30.37  30.73
48  29.18  29.54  29.89
49  28.37  28.71  29.07

## 4
<a id='4'></a>
Find the singular value decomposition of the following matrix:
$$A=
\begin{bmatrix}
    1    &0    &0   \\
    0    &-1   &0   \\
    0    &0    &-2
\end{bmatrix}
$$
Eigenvalues and eigenvectors of $AA^T$are:
$$AA^T = \begin{bmatrix}
    1    &0    &0   \\
    0    &-1   &0   \\
    0    &0    &-2
    \end{bmatrix}
    \begin{bmatrix}
    1    &0    &0   \\
    0    &-1   &0   \\
    0    &0    &-2
    \end{bmatrix}
    =
    \begin{bmatrix}
    1    &0    &0   \\
    0    &1   &0   \\
    0    &0    &4
    \end{bmatrix}
$$
 

$\lambda_1 = 4$, $\lambda_2=1$, $\lambda_3 = 1$, $v_1 = [0,0,1]^T$, $v_2 = [0,1,0]^T$, $v_3 = [1,0,0]^T$

Singular values are 

$\sigma_1 = \sqrt{4} = 2$, 

$\sigma_2 = \sqrt{1} = 1$, 

$\sigma_3 = \sqrt{1} = 1$


$\Sigma$ is a 3-by-3 matrix with the singular values on the diagonal.

$$
\Sigma =
    \begin{bmatrix}
    2    &0    &0   \\
    0    &1   &0   \\
    0    &0    &1
    \end{bmatrix} \\
$$

$$
V = [v_1, v_2, v_3] = \begin{bmatrix}
                        0    &0    &1   \\
                        0    &1    &0   \\
                        1    &0    &0
                        \end{bmatrix}
$$

$$u_1 = \frac{1}{\sigma_1}Av_1 = \begin{bmatrix} 0\\0\\-1\end{bmatrix}, \quad
  u_2 = \frac{1}{\sigma_2}Av_2 = \begin{bmatrix} 0\\-1\\0\end{bmatrix}, \quad
  u_3 = \frac{1}{\sigma_3}Av_3 = \begin{bmatrix} 1\\0\\0\end{bmatrix}$$
  
$$U = \begin{bmatrix}
                        0    &0     &1   \\
                        0    &-1    &0   \\
                        -1    &0    &0
                        \end{bmatrix}
$$

$$A = U\Sigma V^T = \begin{bmatrix}
                        0    &0     &1   \\
                        0    &-1    &0   \\
                        -1    &0    &0
                        \end{bmatrix}
                        \begin{bmatrix}
                        2    &0    &0   \\
                        0    &1   &0   \\
                        0    &0    &1
                        \end{bmatrix}
                        \begin{bmatrix}
                        0    &0    &1   \\
                        0    &1    &0   \\
                        1    &0    &0
                        \end{bmatrix}\\
$$


In [64]:
np.array([[0,0,1],[0,-1,0],[-1,0,0]]).dot(np.array([[2,0,0],[0,1,0],[0,0,1]])).dot(np.array([[0,0,1],[0,1,0],[1,0,0]]))

array([[ 1,  0,  0],
       [ 0, -1,  0],
       [ 0,  0, -2]])

In [81]:
A4 = np.array([[1,0,0],
               [0,-1,0],
               [0,0,-2]
              ])

u, s, v = np.linalg.svd(A4)

print('U =')
u
print('S =')
s
print('V = ')
v

print("Test USV =")
smat = np.zeros(A4.shape)
smat[:len(s), :len(s)] = np.diag(s)

u.dot(smat).dot(v)

np.allclose(A4, np.dot(u, np.dot(smat, v)))

U =


array([[0., 0., 1.],
       [0., 1., 0.],
       [1., 0., 0.]])

S =


array([2., 1., 1.])

V = 


array([[-0., -0., -1.],
       [-0., -1., -0.],
       [ 1.,  0.,  0.]])

Test USV =


array([[ 1.,  0.,  0.],
       [ 0., -1.,  0.],
       [ 0.,  0., -2.]])

True

## 5
<a id='5'></a>
Find the singular value decomposition of the following matrix:

\begin{bmatrix}
    1/2    &1/3    &1/6   \\
    1/4    &1/4    &1/2   \\
    1/6    &1/6    &2/3
\end{bmatrix}

In [82]:
A5 = np.array([[1/2, 1/3, 1/6],
               [1/4, 1/4, 1/2],
               [1/6, 1/6, 2/3]
              ])
u,s,v = np.linalg.svd(A5);

print('U =')
u
print('S =')
s
print('V = ')
v

print("Test USV =")
smat = np.zeros(A5.shape)
smat[:len(s), :len(s)] = np.diag(s)

u.dot(smat).dot(v)

np.allclose(A5, np.dot(u, np.dot(smat, v)))

U =


array([[-0.48789319,  0.83544719,  0.25295892],
       [-0.58962895, -0.10173572, -0.80124125],
       [-0.6436598 , -0.54007206,  0.54224002]])

S =


array([1.03498308, 0.43790388, 0.03064471])

V = 


array([[-0.48177644, -0.40320943, -0.77801903],
       [ 0.69028311,  0.37231105, -0.62039802],
       [ 0.53981542, -0.83594655,  0.09895796]])

Test USV =


array([[0.5       , 0.33333333, 0.16666667],
       [0.25      , 0.25      , 0.5       ],
       [0.16666667, 0.16666667, 0.66666667]])

True

## 6
<a id='6'></a>
Find the singular value decomposition of the following matrix:

\begin{bmatrix}
    5    &4    &2   \\
    3    &2    &1   \\
    8    &6    &3   \\
    1    &2    &1   \\
    0    &2    &1
\end{bmatrix}



In [83]:
A6 = np.array([[5,4,2],
               [3,2,1],
               [8,6,3],
               [1,2,1],
               [0,2,1]
              ])

u,s,v = np.linalg.svd(A6, full_matrices=True)


print('U =')
u
print('S =')
s
print('V = ')
v
print("Test USV =")

smat = np.zeros(A6.shape)
smat[:len(s), :len(s)] = np.diag(s)

u.dot((smat).dot(v))

np.allclose(A6, np.dot(u, np.dot(smat, v)))

U =


array([[-5.07092553e-01,  9.93510118e-16, -3.11652534e-01,
        -5.65464905e-01, -5.70945953e-01],
       [-2.81718085e-01,  1.66666667e-01,  8.99963471e-01,
        -2.84997203e-01,  4.12248528e-02],
       [-7.88810638e-01,  1.66666667e-01, -1.73988843e-01,
         3.87364087e-01,  4.11918617e-01],
       [-1.69030851e-01, -5.00000000e-01,  2.50283005e-01,
         5.83403444e-01, -5.64293728e-01],
       [-1.12687234e-01, -8.33333333e-01, -4.97487723e-03,
        -3.29568690e-01,  4.29204930e-01]])

S =


array([1.32287566e+01, 2.00000000e+00, 1.40980839e-16])

V = 


array([[-0.74535599, -0.59628479, -0.2981424 ],
       [ 0.66666667, -0.66666667, -0.33333333],
       [ 0.        , -0.4472136 ,  0.89442719]])

Test USV =


array([[ 5.0000000e+00,  4.0000000e+00,  2.0000000e+00],
       [ 3.0000000e+00,  2.0000000e+00,  1.0000000e+00],
       [ 8.0000000e+00,  6.0000000e+00,  3.0000000e+00],
       [ 1.0000000e+00,  2.0000000e+00,  1.0000000e+00],
       [-4.4408921e-16,  2.0000000e+00,  1.0000000e+00]])

True

## 7
<a id='7'></a>
Find the singular value decomposition of the following matrix:

\begin{bmatrix}
    1.1    &1      &0.9   \\
    2.4    &2.1    &1.9   \\
    2.5    &2.2    &2.0   \\
    3.1    &2.8    &2.4   \\
    3.3    &3.0    &2.7   \\
    3.9    &3.6    &3.3   \\
\end{bmatrix}

In [84]:
A7 = np.array([[1.1, 1, 0.9],
               [2.4, 2.1, 1.9],
               [2.5, 2.2, 2.0],
               [3.1, 2.8, 2.4],
               [3.3, 3.0, 2.7],
               [3.9, 3.6, 3.3]
              ])
u, s, v = np.linalg.svd(A7)

print('U =')
u
print('S =')
s
print('V = ')
v
print("Test USV =")

smat = np.zeros(A7.shape)
smat[:len(s), :len(s)] = np.diag(s)

u.dot(smat).dot(v)

np.allclose(A7, np.dot(u, np.dot(smat, v)))

U =


array([[-0.15777802, -0.02960049,  0.0266535 , -0.59643203, -0.49347556,
        -0.61177626],
       [-0.33699158,  0.27650762, -0.56781865, -0.46059432, -0.03220267,
         0.52381108],
       [-0.35266165,  0.17686991, -0.55446784,  0.59247072, -0.13712747,
        -0.40876243],
       [-0.43730573,  0.64190424,  0.58088068,  0.13187639, -0.16933014,
         0.11504865],
       [-0.47333405, -0.08880147,  0.07996051, -0.20797038,  0.79073615,
        -0.30522149],
       [-0.56735448, -0.68662772,  0.16006541,  0.14302827, -0.28758401,
         0.2790496 ]])

S =


array([11.01426182,  0.14556223,  0.06962894])

V = 


array([[-0.63502416, -0.57455056, -0.51636806],
       [ 0.63369979, -0.00517288, -0.77356178],
       [-0.44177924,  0.81845274, -0.36737747]])

Test USV =


array([[1.1, 1. , 0.9],
       [2.4, 2.1, 1.9],
       [2.5, 2.2, 2. ],
       [3.1, 2.8, 2.4],
       [3.3, 3. , 2.7],
       [3.9, 3.6, 3.3]])

True

In [69]:
np.allclose?

## 8 
<a id='8'></a>

Suppose that the mortality for ages over 62 is decreasing at a rate of half a p ercent (0.005) a year over the next 40 years. What would the life expectancy coecient be in 2058? Use the table given in https://www.etk.fi/wp-content/uploads/EAK2018.pdf


$$EAL^{2018} = \sum_{x=62}^{n} 1,02^{-(x+0.5-62)} \frac{L_x^{2018}}{l_{62}^{2018}}$$

$$EAK^{2018} = \frac{EAL^{2009}}{EAL^{2018}}$$

$$EAK^{2058} = \frac{EAL^{2009}}{EAL^{2058}}$$


In [70]:
df = pd.read_csv('/home/oscar/Dropbox/StochasticDemography/data/livslangdstabell09-18.csv')

df.head()
df.tail()

age   qx09     lx09     Lx09    val09   qx18     lx18     Lx18    val18
0   62   9.59  1.00000  0.99521  0.98540   8.20  1.00000  0.99590  0.98609
1   63  10.29  0.99041  0.98531  0.95648   8.98  0.99180  0.98735  0.95845
2   64  11.05  0.98022  0.97480  0.92772  10.19  0.98289  0.97789  0.93065
3   65  12.31  0.96939  0.96342  0.89891  11.11  0.97288  0.96747  0.90269
4   66  13.02  0.95745  0.95122  0.87012  11.65  0.96207  0.95647  0.87492

age     qx09     lx09     Lx09    val09     qx18     lx18     Lx18  \
34   96   283.97  0.05830  0.05002  0.02526   277.38  0.07508  0.06467   
35   97   309.15  0.04174  0.03529  0.01747   305.10  0.05425  0.04598   
36   98   332.80  0.02884  0.02404  0.01167   321.87  0.03770  0.03163   
37   99   349.08  0.01924  0.01588  0.00756   340.83  0.02557  0.02121   
38  100  1000.00  0.01252  0.00626  0.00292  1000.00  0.01685  0.00843   

      val18  
34  0.03266  
35  0.02276  
36  0.01535  
37  0.01009  
38  0.00393

In [71]:
#check 16,778288?, 17,458849
np.sum(df['val09'])
np.sum(df['val18'])

16.77827

17.458830000000006

### Adding new $q^{2058}$ values

$$q^{2058}_x = q^{2018}_x \cdot 0.995^{40}$$

In [72]:
r = 0.005
q = 1 - r

df['qx58'] = df['qx18'] * q**40
df.iloc[-1,9] = 1000
df.head()
df.tail()

age   qx09     lx09     Lx09    val09   qx18     lx18     Lx18    val18  \
0   62   9.59  1.00000  0.99521  0.98540   8.20  1.00000  0.99590  0.98609   
1   63  10.29  0.99041  0.98531  0.95648   8.98  0.99180  0.98735  0.95845   
2   64  11.05  0.98022  0.97480  0.92772  10.19  0.98289  0.97789  0.93065   
3   65  12.31  0.96939  0.96342  0.89891  11.11  0.97288  0.96747  0.90269   
4   66  13.02  0.95745  0.95122  0.87012  11.65  0.96207  0.95647  0.87492   

       qx58  
0  6.710225  
1  7.348515  
2  8.338682  
3  9.091537  
4  9.533429

age     qx09     lx09     Lx09    val09     qx18     lx18     Lx18  \
34   96   283.97  0.05830  0.05002  0.02526   277.38  0.07508  0.06467   
35   97   309.15  0.04174  0.03529  0.01747   305.10  0.05425  0.04598   
36   98   332.80  0.02884  0.02404  0.01167   321.87  0.03770  0.03163   
37   99   349.08  0.01924  0.01588  0.00756   340.83  0.02557  0.02121   
38  100  1000.00  0.01252  0.00626  0.00292  1000.00  0.01685  0.00843   

      val18         qx58  
34  0.03266   226.985635  
35  0.02276   249.669469  
36  0.01535   263.392697  
37  0.01009   278.908047  
38  0.00393  1000.000000

### Computing new $l_x^{2058}$
$$l_{62}^{2058} = 1$$
$$l^{2058}_{x+1} = (1 - \frac{q^{2058}_{x}}{1000}) \cdot l_x^{2058}$$

In [73]:
l = 1
l_values = [l]
for q in df['qx58']:
    l = (1 - q / 1000) * l
    l_values.append(l)

df['lx58'] = pd.Series(l_values)
df.head()
df.tail()

age   qx09     lx09     Lx09    val09   qx18     lx18     Lx18    val18  \
0   62   9.59  1.00000  0.99521  0.98540   8.20  1.00000  0.99590  0.98609   
1   63  10.29  0.99041  0.98531  0.95648   8.98  0.99180  0.98735  0.95845   
2   64  11.05  0.98022  0.97480  0.92772  10.19  0.98289  0.97789  0.93065   
3   65  12.31  0.96939  0.96342  0.89891  11.11  0.97288  0.96747  0.90269   
4   66  13.02  0.95745  0.95122  0.87012  11.65  0.96207  0.95647  0.87492   

       qx58      lx58  
0  6.710225  1.000000  
1  7.348515  0.993290  
2  8.338682  0.985991  
3  9.091537  0.977769  
4  9.533429  0.968879

age     qx09     lx09     Lx09    val09     qx18     lx18     Lx18  \
34   96   283.97  0.05830  0.05002  0.02526   277.38  0.07508  0.06467   
35   97   309.15  0.04174  0.03529  0.01747   305.10  0.05425  0.04598   
36   98   332.80  0.02884  0.02404  0.01167   321.87  0.03770  0.03163   
37   99   349.08  0.01924  0.01588  0.00756   340.83  0.02557  0.02121   
38  100  1000.00  0.01252  0.00626  0.00292  1000.00  0.01685  0.00843   

      val18         qx58      lx58  
34  0.03266   226.985635  0.124131  
35  0.02276   249.669469  0.095955  
36  0.01535   263.392697  0.071998  
37  0.01009   278.908047  0.053034  
38  0.00393  1000.000000  0.038243

### Computing new $L_x$ values

$$L_x^{2058} = \frac{l_x^{2058} + l_{x+1}^{2058}}{2}$$

In [74]:
ls = df['lx58'].values

Ls = (ls[:-1] + ls[1:])/2

df['Lx58'] = pd.Series(Ls)
df.iloc[-1, 11] = ls[-1] / 2
df.head()
df.tail()

age   qx09     lx09     Lx09    val09   qx18     lx18     Lx18    val18  \
0   62   9.59  1.00000  0.99521  0.98540   8.20  1.00000  0.99590  0.98609   
1   63  10.29  0.99041  0.98531  0.95648   8.98  0.99180  0.98735  0.95845   
2   64  11.05  0.98022  0.97480  0.92772  10.19  0.98289  0.97789  0.93065   
3   65  12.31  0.96939  0.96342  0.89891  11.11  0.97288  0.96747  0.90269   
4   66  13.02  0.95745  0.95122  0.87012  11.65  0.96207  0.95647  0.87492   

       qx58      lx58      Lx58  
0  6.710225  1.000000  0.996645  
1  7.348515  0.993290  0.989640  
2  8.338682  0.985991  0.981880  
3  9.091537  0.977769  0.973324  
4  9.533429  0.968879  0.964261

age     qx09     lx09     Lx09    val09     qx18     lx18     Lx18  \
34   96   283.97  0.05830  0.05002  0.02526   277.38  0.07508  0.06467   
35   97   309.15  0.04174  0.03529  0.01747   305.10  0.05425  0.04598   
36   98   332.80  0.02884  0.02404  0.01167   321.87  0.03770  0.03163   
37   99   349.08  0.01924  0.01588  0.00756   340.83  0.02557  0.02121   
38  100  1000.00  0.01252  0.00626  0.00292  1000.00  0.01685  0.00843   

      val18         qx58      lx58      Lx58  
34  0.03266   226.985635  0.124131  0.110043  
35  0.02276   249.669469  0.095955  0.083976  
36  0.01535   263.392697  0.071998  0.062516  
37  0.01009   278.908047  0.053034  0.045638  
38  0.00393  1000.000000  0.038243  0.019121

$$EAL^{2058} = \sum_{x=62}^{n} 1,02^{-(x+0.5-62)} \frac{L_x^{2058}}{l_{62}^{2058}}$$

In [75]:
l62 = df.loc[0,'lx58']
vals = []
for age, L in zip(df['age'], df['Lx58']):
    vals.append(1.02 ** -(age + 0.5 - 62) * L)

df['val_58'] = pd.Series(vals) / l62

df.head(20)
df.tail(19)

age   qx09     lx09     Lx09    val09   qx18     lx18     Lx18    val18  \
0    62   9.59  1.00000  0.99521  0.98540   8.20  1.00000  0.99590  0.98609   
1    63  10.29  0.99041  0.98531  0.95648   8.98  0.99180  0.98735  0.95845   
2    64  11.05  0.98022  0.97480  0.92772  10.19  0.98289  0.97789  0.93065   
3    65  12.31  0.96939  0.96342  0.89891  11.11  0.97288  0.96747  0.90269   
4    66  13.02  0.95745  0.95122  0.87012  11.65  0.96207  0.95647  0.87492   
5    67  13.69  0.94499  0.93852  0.84167  12.54  0.95086  0.94490  0.84739   
6    68  15.26  0.93205  0.92494  0.81323  13.68  0.93894  0.93252  0.81989   
7    69  16.54  0.91783  0.91024  0.78461  14.49  0.92609  0.91938  0.79249   
8    70  18.40  0.90265  0.89434  0.75579  16.00  0.91267  0.90537  0.76511   
9    71  20.32  0.88604  0.87704  0.72663  16.75  0.89807  0.89055  0.73783   
10   72  21.98  0.86803  0.85849  0.69733  18.59  0.88303  0.87482  0.71059   
11   73  25.28  0.84895  0.83822  0.66751  21.00  0.86661  0.85751  0.68287   
12   74  26.72  0.82749  0.81644  0.63741  23.26  0.84841  0.83855  0.65467   
13   75  30.65  0.80538  0.79304  0.60701  25.50  0.82868  0.81811  0.62620   
14   76  33.10  0.78070  0.76778  0.57615  28.58  0.80755  0.79601  0.59733   
15   77  38.32  0.75486  0.74039  0.54470  31.30  0.78447  0.77219  0.56810   
16   78  43.19  0.72593  0.71025  0.51228  35.70  0.75991  0.74635  0.53832   
17   79  47.67  0.69458  0.67802  0.47945  39.86  0.73279  0.71818  0.50784   
18   80  53.66  0.66147  0.64372  0.44627  44.53  0.70358  0.68791  0.47690   
19   81  60.03  0.62597  0.60718  0.41268  49.65  0.67225  0.65556  0.44556   

         qx58      lx58      Lx58    val_58  
0    6.710225  1.000000  0.996645  0.986825  
1    7.348515  0.993290  0.989640  0.960676  
2    8.338682  0.985991  0.981880  0.934454  
3    9.091537  0.977769  0.973324  0.908148  
4    9.533429  0.968879  0.964261  0.882051  
5   10.261734  0.959643  0.954719  0.856199  
6   11.194619  0.949795  0.944479  0.830407  
7   11.857459  0.939162  0.933594  0.804742  
8   13.093122  0.928026  0.921951  0.779124  
9   13.706862  0.915876  0.909599  0.753613  
10  15.212571  0.903322  0.896451  0.728156  
11  17.184723  0.889580  0.881936  0.702320  
12  19.034126  0.874293  0.865972  0.676086  
13  20.867163  0.857651  0.848703  0.649611  
14  23.387589  0.839755  0.829935  0.622790  
15  25.613420  0.820115  0.809612  0.595627  
16  29.214028  0.799109  0.787436  0.567953  
17  32.618240  0.775764  0.763112  0.539616  
18  36.439795  0.750460  0.736786  0.510785  
19  40.629594  0.723113  0.708423  0.481492

age     qx09     lx09     Lx09    val09     qx18     lx18     Lx18  \
20   82    66.50  0.58840  0.56883  0.37904    56.48  0.63887  0.62083   
21   83    76.00  0.54927  0.52840  0.34519    64.60  0.60279  0.58332   
22   84    83.82  0.50752  0.48625  0.31143    73.81  0.56385  0.54304   
23   85    94.07  0.46498  0.44311  0.27823    83.45  0.52223  0.50044   
24   86   106.44  0.42124  0.39882  0.24551    97.15  0.47865  0.45540   
25   87   117.49  0.37640  0.35429  0.21382   107.01  0.43215  0.40903   
26   88   130.85  0.33218  0.31045  0.18369   121.88  0.38590  0.36239   
27   89   149.96  0.28871  0.26707  0.15492   135.87  0.33887  0.31585   
28   90   162.49  0.24542  0.22548  0.12823   153.43  0.29283  0.27036   
29   91   188.32  0.20554  0.18619  0.10381   169.50  0.24790  0.22689   
30   92   202.56  0.16683  0.14994  0.08196   190.72  0.20588  0.18625   
31   93   225.06  0.13304  0.11807  0.06327   210.09  0.16661  0.14911   
32   94   238.53  0.10310  0.09080  0.04771   231.33  0.13161  0.11639   
33   95   257.39  0.07851  0.06840  0.03523   257.85  0.10117  0.08812   
34   96   283.97  0.05830  0.05002  0.02526   277.38  0.07508  0.06467   
35   97   309.15  0.04174  0.03529  0.01747   305.10  0.05425  0.04598   
36   98   332.80  0.02884  0.02404  0.01167   321.87  0.03770  0.03163   
37   99   349.08  0.01924  0.01588  0.00756   340.83  0.02557  0.02121   
38  100  1000.00  0.01252  0.00626  0.00292  1000.00  0.01685  0.00843   

      val18         qx58      lx58      Lx58    val_58  
20  0.41368    46.218720  0.693733  0.677701  0.451580  
21  0.38107    52.863480  0.661670  0.644181  0.420827  
22  0.34780    60.400208  0.626692  0.607765  0.389253  
23  0.31423    68.288814  0.588839  0.568734  0.357112  
24  0.28034    79.499800  0.548628  0.526820  0.324308  
25  0.24686    87.568436  0.505012  0.482901  0.291443  
26  0.21442    99.736856  0.460789  0.437810  0.259049  
27  0.18322   111.185155  0.414831  0.391770  0.227262  
28  0.15376   125.554856  0.368708  0.345562  0.196526  
29  0.12651   138.705261  0.322415  0.300055  0.167300  
30  0.10181   156.070013  0.277695  0.256025  0.139951  
31  0.07991   171.920874  0.234355  0.214210  0.114798  
32  0.06115   189.301994  0.194064  0.175696  0.092312  
33  0.04539   211.003843  0.157328  0.140729  0.072490  
34  0.03266   226.985635  0.124131  0.110043  0.055572  
35  0.02276   249.669469  0.095955  0.083976  0.041577  
36  0.01535   263.392697  0.071998  0.062516  0.030345  
37  0.01009   278.908047  0.053034  0.045638  0.021718  
38  0.00393  1000.000000  0.038243  0.019121  0.008921

In [76]:
EAL_58 = np.sum(df['val_58'])

EAL_09 = 16.778288

EAK_58 = EAL_09 / EAL_58

print("EAL 2058 = ", EAL_58)
print("EAK 2058 = ", EAK_58)


EAL 2058 =  18.433020596880397
EAK 2058 =  0.9102299816688512


$$EAK^{2058} = \frac{EAL^{2009}}{EAL^{2058}} = \frac{16.778288}{18.433020596880397} = 0.9102299816688512$$

## 9
<a id='9'></a>

Calculate the dominant eigenvalue λ for the Finnish fertility and mortality data 2015. Assuming the same fertility and mortality over the foreseeable future, what level of immigration would keep the population constant?

Start by reading the mortality data for Finland 1965-2015 and select the data for 2015.

In [77]:
df_mor = pd.read_excel('/home/oscar/Dropbox/StochasticDemography/data/FinlMort65-15.xlsx',index=False)
df_mor = df_mor.iloc[:,:10]
df_mor15 = df_mor[df_mor['year'] == 2015]
df_mor15.head()
df_mor15.tail()

year age_x     m(x)     q(x)  a(x)    l(x)  d(x)   L(x)     T(x)   e(x)
5550  2015     0  0.00168  0.00168  0.15  100000   168  99856  8143956  81.44
5551  2015     1  0.00009  0.00009  0.50   99832     9  99828  8044099  80.58
5552  2015     2  0.00013  0.00013  0.50   99823    13  99817  7944272  79.58
5553  2015     3  0.00008  0.00008  0.50   99810     8  99806  7844455  78.59
5554  2015     4  0.00006  0.00006  0.50   99802     6  99799  7744649  77.60

year age_x     m(x)     q(x)  a(x)  l(x)  d(x)  L(x)  T(x)  e(x)
5656  2015   106  0.68827  0.51205  0.50    50    26    38    71  1.40
5657  2015   107  0.72078  0.52983  0.50    25    13    18    33  1.35
5658  2015   108  0.75118  0.54608  0.50    12     6     8    15  1.30
5659  2015   109  0.77932  0.56080  0.50     5     3     4     7  1.27
5660  2015  110+  0.80513  1.00000  1.24     2     2     3     3  1.24

Then the data for fertality is loaded and the data for year 2015 is selected.

In [78]:
df_fer = pd.read_excel('/home/oscar/Dropbox/StochasticDemography/data/FinlFert39-15.xlsx',index=False, skiprows=2)
df_fer15 = df_fer[df_fer['Year'] == 2015]
df_fer15.head(10)
df_fer15.tail(10)

Year  Age     ASFR
3344  2015  -12  0.00000
3345  2015   13  0.00000
3346  2015   14  0.00021
3347  2015   15  0.00021
3348  2015   16  0.00117
3349  2015   17  0.00412
3350  2015   18  0.00686
3351  2015   19  0.01757
3352  2015   20  0.02768
3353  2015   21  0.03622

Year  Age     ASFR
3378  2015   46  0.00100
3379  2015   47  0.00039
3380  2015   48  0.00025
3381  2015   49  0.00003
3382  2015   50  0.00008
3383  2015   51  0.00000
3384  2015   52  0.00000
3385  2015   53  0.00000
3386  2015   54  0.00000
3387  2015  55+  0.00000

The ASFR-column is the f-values in the Leslie-Matrix, that is the upper row.

In [79]:
upper_row = df_fer15['ASFR'].values
upper_row

array([0.0000e+00, 0.0000e+00, 2.1000e-04, 2.1000e-04, 1.1700e-03,
       4.1200e-03, 6.8600e-03, 1.7570e-02, 2.7680e-02, 3.6220e-02,
       4.6610e-02, 5.4070e-02, 6.5700e-02, 7.8410e-02, 8.9690e-02,
       1.0175e-01, 1.0952e-01, 1.1171e-01, 1.2091e-01, 1.2025e-01,
       1.1069e-01, 1.0393e-01, 9.2840e-02, 8.1970e-02, 6.7980e-02,
       5.7290e-02, 4.2980e-02, 3.5070e-02, 2.4290e-02, 1.5350e-02,
       1.0820e-02, 6.4900e-03, 3.3800e-03, 1.7300e-03, 1.0000e-03,
       3.9000e-04, 2.5000e-04, 3.0000e-05, 8.0000e-05, 0.0000e+00,
       0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00])

The lower diagonal contains the probabilities $s_x$ for survival into the next state.

$$s_x = \frac{l_{x+1}}{l_{x}}$$

In [80]:
L = df_mor15['l(x)'].values
S = L[1:] / L[0:-1]
size = len(L)
size

111

There are as many states as there are ages. So the Leslie-Matrix will be a 111 by 111 matrix.

In [81]:
A = np.zeros([size, size])

A[np.arange(size-1) + 1, np.arange(size-1)] = S
A[0, 12:len(upper_row) + 12] = upper_row

In [82]:
r, c = 8,6
print("Upper left \n{}\n".format(A[:r,: c]))
print("Upper right \n{}\n".format(A[:r, -c:]))

print("Lower left \n{}\n".format(A[-r:, :c]))
print("Lower right \n{}\n".format(A[-r:, -c:]))

print("Upper row \n{}\n".format(A[0, 12:55]))

Upper left 
[[0.         0.         0.         0.         0.         0.        ]
 [0.99832    0.         0.         0.         0.         0.        ]
 [0.         0.99990985 0.         0.         0.         0.        ]
 [0.         0.         0.99986977 0.         0.         0.        ]
 [0.         0.         0.         0.99991985 0.         0.        ]
 [0.         0.         0.         0.         0.99992986 0.        ]
 [0.         0.         0.         0.         0.         0.99991984]
 [0.         0.         0.         0.         0.         0.        ]]

Upper right 
[[0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]]

Lower left 
[[0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0.]]

Lower right 
[[0.         0.         0.         0.         0.        

In [83]:
eig = np.linalg.eigvals(A)

eig_real = [x.real for x in eig if x.imag==0 and x.real !=0]

In [84]:
print("The real valued eigenvalues: ", eig_real)

The real valued eigenvalues:  [-0.8504137534696484, -0.8028890152922405, 1.0160370796313312]


In [85]:
print("All eigenvalues:\n")
eig

All eigenvalues:



array([ 0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,  0.        +0.        j,
        0.        +0.        j,